File: algorithms/04_java/greedy/ActivitySelection.java
import java.util.Arrays;
import java.util.Comparator;

// Activity-selection problem (CLRS 16.1).
public class ActivitySelection {
    static class Activity {
        int start, finish;
        Activity(int s, int f) { start = s; finish = f; }
    }

    public static void main(String[] args) {
        Activity[] acts = {
            new Activity(1, 4), new Activity(3, 5), new Activity(0, 6),
            new Activity(5, 7), new Activity(3, 9), new Activity(5, 9),
            new Activity(6, 10), new Activity(8, 11), new Activity(8, 12),
            new Activity(2, 14), new Activity(12, 16)
        };
        Arrays.sort(acts, Comparator.comparingInt(a -> a.finish));
        int count = 0, last = -1;
        for (Activity a : acts) {
            if (a.start >= last) { count++; last = a.finish; }
        }
        assert count == 4;
        System.out.println("activity selection ok");
    }
}

File: algorithms/02_c/greedy/activity_selection.c
/* Activity-selection problem (CLRS 16.1). */
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct { int start, finish; } Activity;

static int cmp(const void *x, const void *y) {
    const Activity *a = x, *b = y;
    return (a->finish > b->finish) - (a->finish < b->finish);
}

int main(void) {
    Activity acts[] = {{1,4},{3,5},{0,6},{5,7},{3,9},{5,9},{6,10},{8,11},{8,12},{2,14},{12,16}};
    int n = sizeof(acts) / sizeof(acts[0]);
    qsort(acts, n, sizeof(Activity), cmp);
    int count = 0, last = -1;
    for (int i = 0; i < n; ++i)
        if (acts[i].start >= last) { count++; last = acts[i].finish; }
    assert(count == 4);
    printf("activity selection ok\n");
    return 0;
}
